core/arch/riscv/kernel/thread.c

```diff
@@ -103,9 +105,54 @@ short int thread_get_id(void)
 
 short int thread_get_id_may_fail(void)
 {
+	/*
+	 * thread_get_core_local() requires foreign interrupts to be disabled
+	 */
 	uint32_t exceptions = thread_mask_exceptions(THREAD_EXCP_FOREIGN_INTR);
-	short int ct = get_core_pos();
+	struct thread_core_local *l = thread_get_core_local();
+	short int ct = l->curr_thread;
 
 	thread_unmask_exceptions(exceptions);
 	return ct;
 }
+
+static struct thread_core_local *__nostackcheck __unused
+get_core_local(unsigned int pos)
+{
+	/*
+	 * Foreign interrupts must be disabled before playing with core_local
+	 * since we otherwise may be rescheduled to a different core in the
+	 * middle of this function.
+	 */
+	assert(thread_get_exceptions() & THREAD_EXCP_FOREIGN_INTR);
+
+	assert(pos < CFG_TEE_CORE_NB_CORE);
+	return &thread_core_local[pos];
+}
+
+struct thread_core_local *__nostackcheck thread_get_core_local(void)
+{
+	/*
+	 * Rely on thread pointer register instead of hart id which is available
+	 * only in machine mode. This allows retrieving the core local whether
+	 * running core in M-Mode or S-Mode.
+	 */
+	struct thread_core_local *l;
+
+	asm volatile("mv %0, tp" : "=&r"(l));
```

Comment:
We try to avoid using inline assembly in normal functions. Please wrap this in a static inline function you can put in <riscv.h>. A good name could be `read_tp()` if we chose to follow the same convention as in `<arm.h>` and friends.